### PR TITLE
AngularJS + RequireJS example improvements

### PR DESCRIPTION
**Pull request by leog**
As described and discussed on #1276 with @samccone

### CODE REVIEW

**Comment by samccone** [comment]
:+1: this looks good to me @arthurvr want to take a look

**Comment by samccone** [comment]
tests are all passing :shipit:

**Comment by leog** [comment]
Thanks, glad it helped. This pattern really shows when you have a huge AngularJS application, separating pages, directives, providers and controllers in much more manageable pieces of code with clear separation of responsibilities.

Forgot to mention that all tests were passing. And I couldn't help but notice that there is no specific test for the angularjs_require application. I can work on that too using this very same pattern. Afterwards I can take a look at the "sleep" hack to test RequireJS apps, that shouldn't be needed at all.

**Comment by leog** [comment]
Any words about this? Hopefully these changes can help people understand better RequireJS

File: examples/angularjs_require/js/app.js [hunk 1]
```diff
@@ -1,6 +1,18 @@
...
-	return angular.module('todomvc', []);
+require([
+	'angular'
+], function (angular) {
+	require([
+		'controllers/todo', 
+		'directives/todoFocus', 
+		'directives/todoEscape',
+		'services/todoStorage'
+	], function(todoCtrl, todoFocusDir, todoEscapeDir, todoStorageSrv) {
```
Comment (arthurvr) [inline examples/angularjs_require/js/app.js]: code style `function () {}` (the space)

File: examples/angularjs_require/js/controllers/todo.js [hunk 2]
```diff
@@ -7,9 +7,11 @@
...
  */
 
-define(['app', 'services/todoStorage'], function (app) {
-	return app.controller('TodoController', ['$scope', '$location', 'todoStorage', 'filterFilter',
-		function TodoController($scope, $location, todoStorage, filterFilter) {
+define([
+	'angular'
+], function (angular) {
+	return ['$scope', '$location', 'todoStorage', 'filterFilter',
+		function($scope, $location, todoStorage, filterFilter) {
```
Comment (arthurvr) [inline examples/angularjs_require/js/controllers/todo.js]: same thing, code style.

File: examples/angularjs_require/test/package.json [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    "grunt": "0.4.5",
+    "grunt-contrib-jasmine": "0.8.1",
+    "grunt-template-jasmine-requirejs": "0.2.0",
+    "angular-mocks": "^1.3.13",
+    "jquery": "2.1.4"
+  },
+  "scripts": {
+    "test": "grunt jasmine"
+  }
+}
```
Comment (arthurvr) [inline examples/angularjs_require/test/package.json]: Can you cleanup this file a bit? It only needs the `scripts`, the deps and `private`.

File: examples/angularjs_require/test/readme.md [hunk 4]
```diff
@@ -0,0 +1,7 @@
+AngularJS + RequireJS Unit Tests
+================================
+
+To run the test suite, run these commands:
+
+    npm install
+    npm test
```
Comment (arthurvr) [inline examples/angularjs_require/test/readme.md]: This should just go in `examples/angularjs_require/readme.md`.

File: examples/angularjs_require/test/unit/directives/todoEscapeSpec.js [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+	beforeEach(module(todoEscapeDir));
+
+	var triggerKeyDown = function (element, keyCode) {
+		var e = jQuery.Event("keydown");
+		e.keyCode = keyCode;
+		element.triggerHandler(e);
+	};
+
+	describe('todoEscape directive', function() {
```
Comment (arthurvr) [inline examples/angularjs_require/test/unit/directives/todoEscapeSpec.js]: `function () {}`

File: examples/angularjs_require/test/unit/directives/todoEscapeSpec.js [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+			scope = $rootScope.$new();
+			compile = $compile;
+			browser = $browser;
+		}));
+
+		it('should focus on truthy expression', function () {
+			var someValue = false,
+				el = angular.element('<input todo-escape="doSomething()">');
+
+			scope.doSomething = function() {
```
Comment (arthurvr) [inline examples/angularjs_require/test/unit/directives/todoEscapeSpec.js]: `function () {}`

File: examples/angularjs_require/test/unit/directives/todoFocusSpec.js [hunk 7]
```diff
@@ -0,0 +1,32 @@
+define([
+	'directives/todoFocus',
+	'angular-mocks'
+], function (todoFocusDir) {
+	'use strict';
+
+	beforeEach(module(todoFocusDir));
+
+	describe('todoFocus directive', function() {
```
Comment (arthurvr) [inline examples/angularjs_require/test/unit/directives/todoFocusSpec.js]: `function () {}`

Comment (arthurvr) [inline examples/angularjs_require/test/unit/directives/todoFocusSpec.js]: same thing for the other functions ;)

File: examples/angularjs_require/test/gruntfile.js [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+					specs: [
+						'unit/**/*Spec.js'
+					],
+					template: require('grunt-template-jasmine-requirejs'),
+					templateOptions: {
+						requireConfigFile: '../js/main.js',
+						requireConfig: {
+							baseUrl: '../js',
+							paths: {
+								'jquery': '../test/node_modules/jquery/dist/jquery',
```
Comment (arthurvr) [inline examples/angularjs_require/test/gruntfile.js]: Unquote `'jquery'`

File: examples/angularjs_require/test/gruntfile.js [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+								'angular-mocks': '../test/node_modules/angular-mocks/angular-mocks'
+							},
+							shim: {
+								'angular-mocks': ['angular']
+							},
+							deps: ['']
```
Comment (arthurvr) [inline examples/angularjs_require/test/gruntfile.js]: I've never used `grunt-contrib-jasmine` but this seems like strange. What does it do?

**Comment by arthurvr** [comment]
Neat work.

File: examples/angularjs_require/test/gruntfile.js (hunk 9 above)
Comment (leog) [inline examples/angularjs_require/test/gruntfile.js]: It helps generating the ``_SpecRunner.html`` supporting RequireJS modules together with ``grunt-template-jasmine-requirejs``.

**Comment by leog** [comment]
Thanks @arthurvr. I made the changes you suggested.

File: examples/angularjs_require/test/gruntfile.js (hunk 9 above)
Comment (arthurvr) [inline examples/angularjs_require/test/gruntfile.js]: Yeah I knew it does that but don't really understand what's the deal with `deps: ['']`. (The empty string)

**Comment by arthurvr** [comment]
the `test/readme.md` can now be removed isn't it?

File: examples/angularjs_require/test/package.json [hunk 10]
```diff
@@ -0,0 +1,13 @@
+{
+  "private": true,
+  "devDependencies": {
+    "grunt": "0.4.5",
+    "grunt-contrib-jasmine": "0.8.1",
+    "grunt-template-jasmine-requirejs": "0.2.0",
```
Comment (arthurvr) [inline examples/angularjs_require/test/package.json]: Why are these dependencies locked? Seems fine to use them with a caret.

Comment (samccone) [inline examples/angularjs_require/test/package.json]: I am always fine with locking things unless you have infinite trust in the future maintainers of these libs to follow semver.

File: examples/angularjs_require/test/gruntfile.js (hunk 9 above)
Comment (leog) [inline examples/angularjs_require/test/gruntfile.js]: Oh sorry. That is to override the ``main.js`` deps value.

File: examples/angularjs_require/test/package.json (hunk 10 above)
Comment (samccone) [inline examples/angularjs_require/test/package.json]: but yeah best to be consistent and use ^ here as @arthurvr said

Comment (arthurvr) [inline examples/angularjs_require/test/package.json]: That would be inconsistent with what we're doing at the moment. The default assumption should be that they do. Also, yes, I do trust the grunt maintainers lol :)

File: examples/angularjs_require/test/gruntfile.js (hunk 9 above)
Comment (arthurvr) [inline examples/angularjs_require/test/gruntfile.js]: okay. Thanks for the explanation.

**Comment by arthurvr** [comment]
Thanks for all the quick updates @leog. I'm going to go over this this evening once again but looks like I'll merge this one :)

**Comment by leog** [comment]
My pleasure @arthurvr, thanks for taking a look at my suggestions guys :)

File: examples/angularjs_require/test/gruntfile.js (hunk 9 above)
Comment (leog) [inline examples/angularjs_require/test/gruntfile.js]: No problem

**Comment by addyosmani** [comment]
LGTM for a merge. Whenever you're happy with it, @arthurvr :star:

**Comment by arthurvr** [comment]
Super nice work!

**Comment by leog** [comment]
Thanks! :sparkles:

**Pull request closed without merging**